feat: Notifications center and refactor

hellypi opened the pull request.

hellypi committed:
Feat: code refactoring and jest tests

chore: More refactoring

chore: More tests and refactors

fix: Fixed age not getting calculated properly

chore: Implemented mocking in jest

chore: More mock tests

chore: checkStuckScreen tests

RafaelTaranto commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Create tests and rework notifications.js into a notification "module"](https://trello.com/c/T0Oe7HZz/380-create-tests-and-rework-notificationsjs-into-a-notification-module)

hellypi commented:
![](https://github.trello.services/images/mini-trello-icon.png) [More descriptive SMS Transaction notifications](https://trello.com/c/WKTBHWyz/323-more-descriptive-sms-transaction-notifications)

hellypi commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Notifications PT2: create the notifications table schema](https://trello.com/c/XoaXU8wH/426-notifications-pt2-create-the-notifications-table-schema)

hellypi committed:
Feat: make SMS transaction notifications more descriptive

hellypi committed:
Chore: make notifications table migration

hellypi committed:
Feat: save highVolumeTxs on DB and plugins code refactor

Fix: remove unused module and add space before '('

Chore: add jest tests for transactionNotify

hellypi committed:
Feat: save cash balance notifications in DB

Feat: add "detail" row in notifications table migration

This makes searching for specific notifications in the
code like cashbox number easier because we don't need
to scrub the notification message column anymore to search
for notifications relating the cashbox 1 for example.

Feat: clear notifications on cash cassette balance update

hellypi commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Notifications PT4: Cash out fiat alert transactions](https://trello.com/c/66euRo3X/428-notifications-pt4-cash-out-fiat-alert-transactions)

hellypi committed:
Feat: crypto balance notifications saving in DB

Chore: add new column "detail" to transactions table migration

Feat: check if older notification is valid before sending new one

Feat: error saving to database

Fix: fix error when invalidating notification on
clearCryptoBalanceNotifications
Chre: code refactor in new-settings-loader for simplicity
Chore: refactor code on notifier and merge similar functions

hellypi committed:
Feat: compliance blacklisted addresses notifications

liordino reviewed (changes requested):
Squash commits into a single one to make it easier to review (with more commits one could end up reviewing code that is changed on a subsequent commit). Also, try and make some commit messages that are very brief more descriptive.

hellypi committed:
Feat: user compliance saving to DB

hellypi committed:
Chore: make notification center UI

Chore: fiatBalancesNotify refactor

Chore: removed now-unused code in some files

Feat: change column "detail" in database to use jsonb

Chore: add notification center background and button

Chore: notifications screen scaffolding

Fix: change position of notification UI

Feat: join backend and frontend

Feat: notification icons and machine names

Feat: add clear all button, stripe overlay on invalid notification

Fix: rework notification styles

Feat: use popper to render notifications

Feat: make notification center UI

Fix: fix css on notification center

Fix: fix invalidateNotification

Chore: apply PR requested changes

Fix: PR fixes

Fix: make toggleable body/root styles be handled by react

Chore: delete old notifier file

Fix: undo variable name changes for cryptobalance notifs

hellypi commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Notifications PT8: Link the ui to the backend](https://trello.com/c/fJSu9bgO/432-notifications-pt8-link-the-ui-to-the-backend)

chaotixkilla commented on lib/cash-out/cash-out-atomic.js:
```diff
@@ -122,8 +124,9 @@ function updateCassettes (t, tx) {
-  return t.one(sql, values)
-    .then(r => socket.emit(_.assign(r, {op: 'cassetteUpdate', deviceId: tx.deviceId})))
+  return t.one(sql, values).then(r => {
+    return socket.emit(_.assign(r, {op: 'cassetteUpdate', deviceId: tx.deviceId}))
```
{op: 'cassetteUpdate', deviceId: tx.deviceId} could be changed to { op: 'cassetteUpdate', deviceId: tx.deviceId } to follow standard. Confirm with Rafael

Also, isn't the return redundant here?

chaotixkilla commented on lib/customers.js:
```diff
@@ -115,12 +116,21 @@ async function updateCustomer (id, data, userToken) {
 
+const invalidateCustomerNotifications = (id, data) => {
+  if(data.authorized_override === 'verified') {
+    const detailB = notifierUtils.buildDetail({code: 'BLOCKED', customerId: id})
```
StandardJS here as well {code: 'BLOCKED', customerId: id} -> { code: 'BLOCKED', customerId: id }

chaotixkilla commented on lib/machine-loader.js:
```diff
@@ -99,8 +101,9 @@ function renameMachine (rec) {
-  const sql = 'update devices set cassette1=$1, cassette2=$2 where device_id=$3'
-  return db.none(sql, [rec.cassettes[0], rec.cassettes[1], rec.deviceId])
+  const detailB = notifierUtils.buildDetail({deviceId: rec.deviceId})
+  const sql = `update devices set cassette1=$1, cassette2=$2 where device_id=$3;`
```
Normalize SQL query to uppercase here

chaotixkilla commented on lib/notifier/email.js:
```diff
@@ -0,0 +1,102 @@
+  let body = 'Errors were reported by your Lamassu Machines.\n'
+
+  if (config.balance && alertRec.general.length !== 0) {
+    body = body + '\nGeneral errors:\n'
+    body = body + emailAlerts(alertRec.general) + '\n'
+  }
+
+  _.keys(alertRec.devices).forEach(function (device) {
+    const deviceName = alertRec.deviceNames[device]
+    body = body + '\nErrors for ' + deviceName + ':\n'
+
+    let alerts = []
+    if (config.balance) {
+      alerts = _.concat(alerts, alertRec.devices[device].balanceAlerts)
+    }
+
+    if (config.errors) {
+      alerts = _.concat(alerts, alertRec.devices[device].deviceAlerts)
+    }
+
+    body = body + emailAlerts(alerts)
```
Can replace all these 'body = body + ...' for 'body += ...'

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+      )
+      if (!currentAlertFingerprint) {
+        const inAlert = !!utils.getAlertFingerprint()
+        // (fingerprint = null, lastAlertTime = null)
```
I believe this is an information annotation, but pointing it out just in case

chaotixkilla commented on new-lamassu-admin/src/styling/icons/action/wrench/zodiac.svg:
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<svg width="32px" height="32px" viewBox="0 0 24 24" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
+    <title>small-wrench</title>
```
These titles might appear on mouse hovering, which is not indented most likely

chaotixkilla commented on new-lamassu-admin/src/styling/icons/arrow/transaction.svg:
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<svg width="24px" height="24px" viewBox="0 0 18 18" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
+    <title>transaction</title>
```
Same here

chaotixkilla commented on new-lamassu-admin/src/pages/Customers/components/CustomerDetails.js:
```diff
@@ -47,9 +47,8 @@ const CustomerDetails = memo(({ customer, locale, setShowCompliance }) => {
-              ? name
-              : getFormattedPhone(R.path(['phone'])(customer), locale.country)}
+            {name.length ? name : R.path(['phone'])(customer)}
+            {/* getFormattedPhone(R.path(['phone'])(customer), locale.country)} */}
```
Remove annotation

chaotixkilla commented on new-lamassu-admin/src/pages/Customers/components/CustomerDetails.js:
```diff
@@ -22,7 +22,7 @@ const CustomerDetails = memo(({ customer, locale, setShowCompliance }) => {
       header: 'Phone number',
       size: 172,
-      value: getFormattedPhone(customer.phone, locale.country)
+      value: customer.phone // getFormattedPhone(customer.phone, locale.country)
```
Remove annotation

chaotixkilla commented on new-lamassu-admin/src/pages/Customers/components/CustomerDetails.js:
```diff
@@ -9,7 +9,7 @@ import { ReactComponent as LawIconInverse } from 'src/styling/icons/circle butto
 
 import mainStyles from '../CustomersList.styles'
-import { getFormattedPhone, getName } from '../helper'
+import { /* getFormattedPhone, */ getName } from '../helper'
```
Remove annotation

chaotixkilla commented on new-lamassu-admin/src/pages/Customers/CustomersList.js:
```diff
@@ -22,7 +26,7 @@ const CustomersList = ({ data, locale, onClick, loading }) => {
       header: 'Phone',
       width: 172,
-      view: it => getFormattedPhone(it.phone, locale.country)
+      view: it => it.phone // getFormattedPhone(it.phone, locale.country)
```
Remove annotation

chaotixkilla commented on new-lamassu-admin/src/pages/Customers/CustomersList.js:
```diff
@@ -11,7 +11,11 @@ import { ReactComponent as TxOutIcon } from 'src/styling/icons/direction/cash-ou
+import {
+  getAuthorizedStatus,
+  getName
+  /* getFormattedPhone */
```
Remove annotation

chaotixkilla commented on new-lamassu-admin/src/pages/Customers/CustomerProfile.js:
```diff
@@ -147,12 +147,13 @@ const CustomerProfile = memo(() => {
-          {name.length
+          {name.length ? name : R.path(['phone'])(customerData)}
+          {/*           {name.length
             ? name
             : getFormattedPhone(
                 R.path(['phone'])(customerData),
                 locale.country
-              )}
+              )} */}
```
Remove annotation

chaotixkilla commented on new-lamassu-admin/src/pages/Customers/CustomerProfile.js:
```diff
@@ -24,7 +24,7 @@ import {
   ComplianceDetails
 } from './components'
-import { getFormattedPhone, getName } from './helper'
+import { /*  getFormattedPhone,  */ getName } from './helper'
```
Remove annotation

RafaelTaranto replied:
same thing

chaotixkilla commented on new-lamassu-admin/src/components/layout/Header.js:
```diff
@@ -56,8 +70,31 @@ const Header = memo(({ tree }) => {
+  const id = popperOpen ? 'notifications-popper' : undefined
   return (
-    <header>
+    <header style={{ position: 'relative' }}>
```
Move the styling to the stylesheet if possible

chaotixkilla commented on new-lamassu-admin/src/components/layout/Header.js:
```diff
@@ -85,14 +122,41 @@ const Header = memo(({ tree }) => {
+                id={id}
+                open={popperOpen}
+                anchorEl={anchorEl}
+                style={{ zIndex: 1 }}
```
Can this styling be moved to the stylesheet?

chaotixkilla commented on new-lamassu-admin/src/components/NotificationCenter/NotificationRow.js:
```diff
@@ -0,0 +1,82 @@
+          </Label2>
+        </Grid>
+        <Grid item xs={12}>
+          <TL2 style={{ ...styles.notificationBody }}>{message}</TL2>
+        </Grid>
+        <Grid item xs={12}>
+          <Label1 className={classes.notificationSubtitle}>
+            {buildAge(created)}
+          </Label1>
+        </Grid>
+      </Grid>
+      <Grid item xs={3} style={{ zIndex: 1 }}>
```
Move styles to the stylesheet

chaotixkilla commented on new-lamassu-admin/src/components/NotificationCenter/NotificationCenter.js:
```diff
@@ -0,0 +1,143 @@
+              color="primary"
+              Icon={ShowUnreadIcon}
+              InverseIcon={ClearAllIconInverse}
+              style={{ ...styles.clearAllButton }}
+              onClick={() => setShowingUnread(!showingUnread)}>
+              {showingUnread ? 'Show all' : 'Show unread'}
+            </ActionButton>
+          )}
+          <ActionButton
+            color="primary"
+            Icon={ClearAllIcon}
+            InverseIcon={ClearAllIconInverse}
+            style={{ ...styles.clearAllButton }}
```
To avoid inline styling and keep it all in the stylesheet, check the changes I made to the ActionButton container on another PR

https://github.com/lamassu/lamassu-server/pull/537/files#diff-38b5504c808007220c5af0ac5dcab3170817148adcf9966a8605527f08dfcb74

chaotixkilla commented on lib/routes.js:
```diff
@@ -296,6 +296,7 @@ function updateCustomer (req, res, next) {
 }
 
 function triggerSanctions (req, res, next) {
+  console.log("SANCTIONS TRIGGERED")
```
A forgotten console.log here

chaotixkilla commented on lib/plugins/sms/mock-sms/mock-sms.js:
```diff
@@ -2,7 +2,7 @@ const _ = require('lodash/fp')
 exports.NAME = 'MockSMS'
 
-exports.sendMessage = function sendMessage (account, rec) {
+exports.sendMessage = function sendMessage(account, rec) {
```
Removed spacing here for no reason

chaotixkilla commented on lib/plugins.js:
```diff
@@ -262,11 +297,18 @@ function plugins (settings, deviceId) {
-      db.none(pgp.helpers.update(devices, null, 'devices') + 'WHERE device_id = ${deviceId}', {
-        deviceId
-      })
+      db.none(
+        `insert into machine_pings(device_id, device_time) values($1, $2) 
+            ON CONFLICT (device_id) DO UPDATE SET device_time = $2, updated = now()`,
+        [deviceId, deviceTime]
+      ),
+      db.none(
+        pgp.helpers.update(devices, null, 'devices') +
+          'WHERE device_id = ${deviceId}',
+        {
+          deviceId
+        }
+      )
```
Not really the point of this PR, but missing standardized uppercase SQL command

chaotixkilla commented on lib/plugins.js:
```diff
@@ -340,99 +385,31 @@ function plugins (settings, deviceId) {
-        const sql = 'update cash_out_txs set notified=$1 where id=$2'
-        const values = [true, tx.id]
+    return sms.sendMessage(settings, rec).then(() => {
+      const sql = 'update cash_out_txs set notified=$1 where id=$2'
```
Same here

chaotixkilla reviewed (changes requested):
Mostly a preliminary review, but some minor stuff around some files.

chaotixkilla commented on lib/cash-in/cash-in-tx.js:
```diff
@@ -8,6 +8,7 @@ const plugins = require('../plugins')
 const logger = require('../logger')
 const settingsLoader = require('../new-settings-loader')
 const configManager = require('../new-config-manager')
+const notifier = require("../notifier/index")
```
Same applies here and you're using double quotes on the import, instead of single quotes.

```suggestion
const notifier = require('../notifier')
```

chaotixkilla commented on lib/cash-in/cash-in-tx.js:
```diff
@@ -30,10 +31,11 @@ function post (machineTx, pi) {
           if (_.some(it => it.created_by_operator === true)(blacklistItems)) {
             blacklisted = true
+            notifier.blacklistNotify(r.tx, false)
           } else if (_.some(it => it.created_by_operator === false)(blacklistItems) && rejectAddressReuseActive) {
```
I know this is existing code, but can't we simplify the check using (it => it.created_by_operator) and (it => !created_by_operator)?

```suggestion
          if (_.some(it => it.created_by_operator)(blacklistItems)) {
            blacklisted = true
            notifier.blacklistNotify(r.tx, false)
          } else if (_.some(it => !it.created_by_operator)(blacklistItems) && rejectAddressReuseActive) {
```

chaotixkilla commented on lib/cash-out/cash-out-atomic.js:
```diff
@@ -9,6 +9,8 @@ const helper = require('./cash-out-helper')
 const cashOutActions = require('./cash-out-actions')
 const cashOutLow = require('./cash-out-low')
 
+const notifier = require("../notifier/index")
```
Use single quotes here as well

```suggestion
const notifier = require('../notifier')
```

chaotixkilla commented on lib/machine-loader.js:
```diff
@@ -83,28 +85,29 @@ function getMachineNames (config) {
 function resetCashOutBills (rec) {
-  const sql = 'update devices set cassette1=$1, cassette2=$2 where device_id=$3'
-  return db.none(sql, [rec.cassettes[0], rec.cassettes[1], rec.deviceId])
+  const detailB = notifierUtils.buildDetail({deviceId: rec.deviceId})
```
Format like standardJS

```suggestion
  const detailB = notifierUtils.buildDetail({ deviceId: rec.deviceId })
```

chaotixkilla commented on lib/notifier/email.js:
```diff
@@ -0,0 +1,102 @@
+const _ = require('lodash/fp')
+const utils = require("./utils")
```
Single quotes here

```suggestion
const utils = require('./utils')
```

chaotixkilla commented on lib/notifier/email.js:
```diff
@@ -0,0 +1,102 @@
+  LOW_CASH_OUT
+} = require('./codes')
+
+function alertSubject(alertRec, config) {
```
Missing space before the argument list

```suggestion
function alertSubject (alertRec, config) {
```

chaotixkilla commented on lib/notifier/email.js:
```diff
@@ -0,0 +1,102 @@
+      alerts = _.concat(alerts, alertRec.devices[device].deviceAlerts)
+    }
+  })
+
```
Shouldn't this be like the suggestion? To avoid using the default forEach function

```suggestion
  _.forEach(_.keys(alertRec.devices), (device) => {
    if (config.balance)
      alerts = _.concat(alerts, alertRec.devices[device].balanceAlerts)

    if (config.errors)
      alerts = _.concat(alerts, alertRec.devices[device].deviceAlerts)
})
```

chaotixkilla commented on lib/notifier/email.js:
```diff
@@ -0,0 +1,102 @@
+
+  if (alerts.length === 0) return null
+
+  const alertTypes = _.map(codeDisplay, _.uniq(_.map('code', alerts))).sort()
```
Using default JS lib `sort()` function. Consider lodash's `sortBy()` with `sortedUniq()` or something similar.

chaotixkilla commented on lib/notifier/email.js:
```diff
@@ -0,0 +1,102 @@
+  return '[Lamassu] Errors reported: ' + alertTypes.join(', ')
+}
+
+function printEmailAlerts(alertRec, config) {
```
```suggestion
function printEmailAlerts (alertRec, config) {
```

chaotixkilla commented on lib/notifier/email.js:
```diff
@@ -0,0 +1,102 @@
+
+    if (config.errors) {
+      alerts = _.concat(alerts, alertRec.devices[device].deviceAlerts)
+    }
```
Same issues as above, use `_.forEach()`. Also, consider a Pull Up Method refactoring somewhere on this section of code, as the functionality seems to be similar to the snippet a bit above

```suggestion
  _.forEach(_.keys(alertRec.devices), (device) => {
    const deviceName = alertRec.deviceNames[device]
    body += '\nErrors for ' + deviceName + ':\n'
    let alerts = []
	
    if (config.balance)
      alerts = _.concat(alerts, alertRec.devices[device].balanceAlerts)

    if (config.errors)
      alerts = _.concat(alerts, alertRec.devices[device].deviceAlerts)
  })
```

chaotixkilla commented on lib/notifier/email.js:
```diff
@@ -0,0 +1,102 @@
+  return alerts.map(emailAlert).join('\n') + '\n'
+}
+
+function emailAlert(alert) {
```
```suggestion
function emailAlert (alert) {
```

chaotixkilla commented on lib/notifier/email.js:
```diff
@@ -0,0 +1,102 @@
+  return body
+}
+
+function emailAlerts(alerts) {
+  return alerts.map(emailAlert).join('\n') + '\n'
+}
```
```suggestion
function emailAlerts (alerts) {
  return _.join(_.map(alerts, emailAlert), '\n') + '\n'
}
```

hellypi replied:
lodash/fp has the arguments the other way around, but good suggestion

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+const emailFuncs = require('./email')
+const smsFuncs = require('./sms')
+const { STALE, STALE_STATE, PING } = require('./codes')
+const {NOTIFICATION_TYPES: {HIGH_VALUE_TX, FIAT_BALANCE, CRYPTO_BALANCE, COMPLIANCE, ERROR}} = require('./codes')
```
```suggestion
const {
  NOTIFICATION_TYPES: {
    HIGH_VALUE_TX,
    FIAT_BALANCE,
    CRYPTO_BALANCE,
    COMPLIANCE,
    ERROR
  }
} = require('./codes')
```

Not sure if my formatting is correct, but I guess the idea passes through

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+const { STALE, STALE_STATE, PING } = require('./codes')
+const {NOTIFICATION_TYPES: {HIGH_VALUE_TX, FIAT_BALANCE, CRYPTO_BALANCE, COMPLIANCE, ERROR}} = require('./codes')
+
+function buildMessage(alerts, notifications) {
```
```suggestion
function buildMessage (alerts, notifications) {
```

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+  return rec
+}
+
+function checkNotification(plugins) {
```
```suggestion
function checkNotification (plugins) {
```

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+      }
+      if (utils.shouldNotAlert(currentAlertFingerprint)) {
+        return
+      }
```
```suggestion
        if (inAlert) return utils.sendNoAlerts(plugins, smsEnabled, emailEnabled)
      }
      if (utils.shouldNotAlert(currentAlertFingerprint)) return
```

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+    .catch(logger.error)
+}
+
+function getAlerts(plugins) {
```
```suggestion
function getAlerts (plugins) {
```

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+  })
+}
+
+function buildAlerts(pings, balances, events, devices) {
```
```suggestion
function buildAlerts (pings, balances, events, devices) {
```

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+    alerts.devices[deviceId].deviceAlerts = _.isEmpty(ping) ? stuckScreen : ping
+
+    alerts.deviceNames[deviceId] = deviceName
+  })
```
```suggestion
  _.forEach(devices, function (device) {
    const deviceId = device.deviceId
    const deviceName = device.name
    const deviceEvents = events.filter(function (eventRow) {
      return eventRow.device_id === deviceId
    })
    const ping = pings[deviceId] || []
    const stuckScreen = checkStuckScreen(deviceEvents, deviceName)

    if (!alerts.devices[deviceId]) alerts.devices[deviceId] = {}
    alerts.devices[deviceId].balanceAlerts = _.filter(
      ['deviceId', deviceId],
      balances
    )
    alerts.devices[deviceId].deviceAlerts = _.isEmpty(ping) ? stuckScreen : ping

    alerts.deviceNames[deviceId] = deviceName
  })
```

Use lodash here

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+  return alerts
+}
+
+function checkPings(devices) {
```
```suggestion
function checkPings (devices) {
```

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+  return _.zipObject(deviceIds)(pings)
+}
+
+function checkStuckScreen(deviceEvents, machineName) {
```
```suggestion
function checkStuckScreen (deviceEvents, machineName) {
```

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+
+  if (!lastEvent) {
+    return []
+  }
```
```suggestion
  if (!lastEvent) return []
```

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+
+  if (isIdle) {
+    return []
+  }
```
```suggestion
  if (isIdle) return []
```

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+  const age = Math.floor(lastEvent.age)
+  if (age > STALE_STATE) {
+    return [{ code: STALE, state, age, machineName }]
+  }
```
```suggestion
  if (age > STALE_STATE) return [{ code: STALE, state, age, machineName }]
```

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+  const isCashOut = tx.direction === 'cashOut'
+  
+  // high value tx on database
+  if(highValueTx && tx.direction === 'cashIn' || highValueTx && tx.direction === 'cashOut' && rec.isRedemption) {
```
Consider simplifying this condition, it is very messy. For example, you're repeating `highValueTx` on both sides and the lack of parenthesis makes it really hard to understand what the code is trying to check

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+  if(highValueTx && tx.direction === 'cashIn' || highValueTx && tx.direction === 'cashOut' && rec.isRedemption) {
+    const direction = tx.direction === "cashOut" ? 'cash-out' : 'cash-in'
+    const message = `${tx.fiat} ${tx.fiatCode} ${direction} transaction`
+    const detailB = utils.buildDetail({deviceId: tx.deviceId, direction, fiat: tx.fiat, fiatCode: tx.fiatCode, cryptoAddress: tx.toAddress})
```
```suggestion
    const detailB = utils.buildDetail({ deviceId: tx.deviceId, direction, fiat: tx.fiat, fiatCode: tx.fiatCode, cryptoAddress: tx.toAddress })
```

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+  
+  if (!notificationsEnabled && !highValueTx) return Promise.resolve()
+  if (zeroConf && isCashOut && !rec.isRedemption && !rec.error) return Promise.resolve()
+  if (!zeroConf && rec.isRedemption) return sendRedemptionMessage(tx.id, rec.error)
```
You can assign more readable statements for those conditions, especially the middle if statement

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+  .then(([msg, highValueTx]) => sendTransactionMessage(msg, highValueTx))
+}
+
+function sendRedemptionMessage(txId, error) {
```
```suggestion
function sendRedemptionMessage (txId, error) {
```

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+  return sendTransactionMessage(rec)
+}
+
+async function sendTransactionMessage(rec, isHighValueTx) {
```
```suggestion
async function sendTransactionMessage (rec, isHighValueTx) {
```

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+  const smsActive =
+    notifications.sms.active &&
+    (notifications.sms.transactions || isHighValueTx)
+  if (smsActive) promises.push(smsFuncs.sendMessage(settings, rec))
```
If you are using `.push()` on an array, you can initialize it with `const`. That prevents reassignment of the array, but allows for push and pop operations to occur. See snippet on image 
![image](https://user-images.githubusercontent.com/18738838/104453717-bc755f80-559c-11eb-8269-725ab752262c.png)


```suggestion
  const promises = []

  const emailActive =
    notifications.email.active &&
    (notifications.email.transactions || isHighValueTx)
  if (emailActive) promises.push(emailFuncs.sendMessage(settings, rec))

  const smsActive =
    notifications.sms.active &&
    (notifications.sms.transactions || isHighValueTx)
  if (smsActive) promises.push(smsFuncs.sendMessage(settings, rec))
```

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+      else {
+        // if it exists then it is a duplicate, add it to array
+        duplicateIndexes.push(idx)
+      }
```
You are applying it here, so I guess it was forgotten haha 😄

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+    return cryptoWarnings.forEach((balance, idx) => {
+      if(duplicateIndexes.includes(idx)) {
+        return
+      }
```
```suggestion
      if(duplicateIndexes.includes(idx)) return
```

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+      }
+      const fiat = utils.formatCurrency(balance.fiatBalance.balance, balance.fiatCode)
+      const message = `${balance.code === 'HIGH_CRYPTO_BALANCE' ? 'High' : 'Low'} balance in ${balance.cryptoCode} [${fiat}]`
+      console.log(`Adding ${balance.code === 'HIGH_CRYPTO_BALANCE' ? 'high' : 'low'} balance notification for ${balance.cryptoCode}`)
```
Forgotten console.log?

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+      else {
+        // if it exists then it is a duplicate, add it to array
+        duplicateIndexes.push(idx)
+      }
```
```suggestion
      idx === -1 ? idsToInvalidate.push(notification.id) : duplicateIndexes.push(idx)
```

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+  return queries.getAllValidNotifications(FIAT_BALANCE).then(notifications => {
+    const duplicateIndexes = []
+    const idsToInvalidate = []
+    _.forEach(notification => {
```
Hmm, TIL. Thought lodash's `forEach()` took the collection as first argument and not the iteration function

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+    return fiatWarnings.forEach((balance, idx) => {
+      if(duplicateIndexes.includes(idx)) {
+        return
+      }
```
```suggestion
      if(duplicateIndexes.includes(idx)) return
```

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+      }, alerts)
+      if(idx !== -1) {
+        return
+      }
```
```suggestion
      if(idx !== -1) return
```

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+    alerts = _.concat(alerts, alertRec.devices[device].deviceAlerts)
+  })
+  
+  //    now that we have all the alerts, we want to add PING and STALE alerts to the DB
+  //    if there is a valid alert on the DB that doesn't exist on the new alerts array,
+  // that alert should be considered invalid
+  //    after that, for the alerts array, we have to see if there is a valid alert of 
+  // the sorts already on the DB
```
It's just a little nitpick of mine, and you can change if you want, as it is not really an issue, per say. I noticed this also happened somewhere above this section, but, when you need to be so descriptive on a comment, that most likely is an indicator that the code is not clear enough for someone to read in the future and understand what is going on

hellypi replied:
that was a comment to help me write the code rather than explain what it does, really. removed

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+          return queries.getValidNotifications(ERROR, _.omit(['age'], detailB)).then(res => {
+            if(res.length > 0) {
+              return Promise.resolve()
+            }
```
```suggestion
            if(res.length > 0)  return Promise.resolve()
```

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+          return queries.getValidNotifications(ERROR, detailB).then(res => {
+            if(res.length > 0) {
+              return Promise.resolve()
+            }
```
```suggestion
            if(res.length > 0) return Promise.resolve()
```

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+  // code for now can be "BLOCKED", "SUSPENDED"
+  const detailB = utils.buildDetail({customerId: customer.id, code, deviceId})
+  const date = new Date()
+  if (days) {
+    date.setDate(date.getDate() + days)
+  }
```
```suggestion
  if (days) date.setDate(date.getDate() + days)
```

hellypi replied:
I confirmed with Rafael that non-returning code can be inside brackets even if it's only one line

chaotixkilla commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,424 @@
+  }).then(res => {
+    if (res.length > 0) {
+      return Promise.resolve()
+    }
```
```suggestion
    if (res.length > 0) return Promise.resolve()
```

chaotixkilla commented on lib/blacklist.js:
```diff
@@ -1,4 +1,5 @@
 const db = require('./db')
+const notifier = require('./notifier/index')
```
I believe that, if you omit the `/index`, it knows that's the entrypoint by default

```suggestion
const notifier = require('./notifier')
```

chaotixkilla commented on lib/notifier/queries.js:
```diff
@@ -0,0 +1,81 @@
+*/
+
+const addNotification = (type, message, detail) => {
+    const sql = `INSERT INTO notifications (id, type, message, detail) values ($1, $2, $3, $4)`
```
```suggestion
    const sql = `INSERT INTO notifications (id, type, message, detail) VALUES ($1, $2, $3, $4)`
```

chaotixkilla commented on lib/notifier/sms.js:
```diff
@@ -0,0 +1,67 @@
+const utils = require('./utils')
+const sms = require('../sms')
+
+function printSmsAlerts(alertRec, config) {
```
```suggestion
function printSmsAlerts (alertRec, config) {
```

chaotixkilla commented on lib/notifier/sms.js:
```diff
@@ -0,0 +1,67 @@
+function printSmsAlerts(alertRec, config) {
+  let alerts = []
+
+  if (config.balance) {
+    alerts = _.concat(alerts, alertRec.general)
+  }
```
```suggestion
  if (config.balance) alerts = _.concat(alerts, alertRec.general)
```

chaotixkilla commented on lib/notifier/sms.js:
```diff
@@ -0,0 +1,67 @@
+    alerts = _.concat(alerts, alertRec.general)
+  }
+
+  _.keys(alertRec.devices).forEach(function (device) {
+    if (config.balance) {
+      alerts = _.concat(alerts, alertRec.devices[device].balanceAlerts)
+    }
+
+    if (config.errors) {
+      alerts = _.concat(alerts, alertRec.devices[device].deviceAlerts)
+    }
+  })
```
Yeah, the same logic is now applied on three different places. Consider the Pull Up Method refactoring

```suggestion
  _.forEach(_.keys(alertRec.devices), (device) => {	
    if (config.balance)
      alerts = _.concat(alerts, alertRec.devices[device].balanceAlerts)

    if (config.errors)
      alerts = _.concat(alerts, alertRec.devices[device].deviceAlerts)
  })
```

chaotixkilla commented on lib/notifier/sms.js:
```diff
@@ -0,0 +1,67 @@
+    }
+  }, _.toPairs(alertsMap))
+
+  const mapByCodeDisplay = _.map(it => {
+    if(_.isEmpty(it.machineNames) && _.isEmpty(it.cryptoCodes)) {
+      return it.codeDisplay
+    }
+    if(_.isEmpty(it.machineNames)) {
+      return `${it.codeDisplay} (${it.cryptoCodes.join(', ')})`
+    }
+    else return `${it.codeDisplay} (${it.machineNames.join(', ')})`
+  })
```
```suggestion
  const mapByCodeDisplay = _.map(it => {
    if(_.isEmpty(it.machineNames) && _.isEmpty(it.cryptoCodes)) return it.codeDisplay
    if(_.isEmpty(it.machineNames)) return `${it.codeDisplay} (${it.cryptoCodes.join(', ')})`
    return `${it.codeDisplay} (${it.machineNames.join(', ')})`
  })
```

chaotixkilla commented on lib/notifier/test/notifier.test.js:
```diff
@@ -0,0 +1,337 @@
+
+const notifier = require('..')
+const utils = require('../utils')
+const queries = require("../queries")
```
```suggestion
const queries = require('../queries')
```

chaotixkilla commented on lib/notifier/utils.js:
```diff
@@ -0,0 +1,195 @@
+}
+
+function checkPing(device) {
+  const age = +Date.now() - +new Date(device.lastPing)
```
Don't really understand what is going on with those `-` and `+` operators

chaotixkilla commented on lib/notifier/utils.js:
```diff
@@ -0,0 +1,195 @@
+  let alerts = []
+  if (!isActive(config)) return alerts
+
+  if (config.balance) {
+    alerts = _.concat(alerts, alertRec.general)
+  }
```
```suggestion
  if (config.balance) alerts = _.concat(alerts, alertRec.general)
```

chaotixkilla commented on lib/notifier/utils.js:
```diff
@@ -0,0 +1,195 @@
+    alerts = _.concat(alerts, alertRec.general)
+  }
+
+  _.keys(alertRec.devices).forEach(function (device) {
+    if (config.balance) {
+      alerts = _.concat(alerts, alertRec.devices[device].balanceAlerts)
+    }
+
+    if (config.errors) {
+      alerts = _.concat(alerts, alertRec.devices[device].deviceAlerts)
+    }
+  })
```
Once again here, logic being repeated
```suggestion
  _.forEach(_.keys(alertRec.devices), (device) => {	
    if (config.balance)
      alerts = _.concat(alerts, alertRec.devices[device].balanceAlerts)

    if (config.errors)
      alerts = _.concat(alerts, alertRec.devices[device].deviceAlerts)
  })
```

chaotixkilla commented on lib/notifier/utils.js:
```diff
@@ -0,0 +1,195 @@
+  )
+}
+
+function getAlertTypes(alertRec, config) {
```
```suggestion
function getAlertTypes (alertRec, config) {
```

chaotixkilla commented on lib/notifier/utils.js:
```diff
@@ -0,0 +1,195 @@
+  return alerts
+}
+
+function buildAlertFingerprint(alertRec, notifications) {
```
```suggestion
function buildAlertFingerprint (alertRec, notifications) {
```

chaotixkilla commented on lib/notifier/utils.js:
```diff
@@ -0,0 +1,195 @@
+  return crypto.createHash('sha256').update(subject).digest('hex')
+}
+
+function sendNoAlerts(plugins, smsEnabled, emailEnabled) {
```
```suggestion
function sendNoAlerts (plugins, smsEnabled, emailEnabled) {
```

chaotixkilla commented on lib/notifier/utils.js:
```diff
@@ -0,0 +1,195 @@
+  const subject = '[Lamassu] All clear'
+
+  let rec = {}
+  if (smsEnabled) {
+    rec = _.set(['sms', 'body'])(subject)(rec)
+  }
```
```suggestion
  if (smsEnabled) rec = _.set(['sms', 'body'])(subject)(rec)
```

chaotixkilla commented on lib/notifier/utils.js:
```diff
@@ -0,0 +1,195 @@
+  const customerName = customer.name || customer.id
+  const phone = customer.phone ? `- Phone: ${customer.phone}` : ''
+
+  let status = null
+  if (rec.error) {
+    status = `Error - ${rec.error}`
+  } else {
+    status = !isCashOut
+      ? 'Successful'
+      : !rec.isRedemption
+      ? 'Successful & awaiting redemption'
+      : 'Successful & dispensed'
+  }
```
Is there a way to simplify that ternary inside a ternary?

chaotixkilla commented on lib/notifier/utils.js:
```diff
@@ -0,0 +1,195 @@
+  }, highValueTx]
+}
+
+function formatCurrency(num, code) {
```
```suggestion
function formatCurrency (num, code) {
```

chaotixkilla commented on lib/plugins.js:
```diff
@@ -22,6 +21,8 @@ const machineLoader = require('./machine-loader')
 const customers = require('./customers')
 const coinUtils = require('./coin-utils')
 
+const notifier = require('./notifier/index')
```
```suggestion
const notifier = require('./notifier')
```

RafaelTaranto replied:
I've written this above, but kinda got lost. We don't really need to keep the old notifier file, so get rid of it and update the imports.

chaotixkilla commented on lib/poller.js:
```diff
@@ -1,7 +1,7 @@
 
 const plugins = require('./plugins')
-const notifier = require('./notifier')
+const notifier = require('./notifier/index')
```
```suggestion
const notifier = require('./notifier')
```

hellypi replied:
there are two notifier files, the one outside the notifier folder is the pre-refactor code. which is why I explicitly import the index.js file, to grab the new code.

RafaelTaranto replied:
Why not remove the old one? We don't want to keep unused files around.

chaotixkilla commented on lib/routes.js:
```diff
@@ -25,6 +25,7 @@ const E = require('./error')
 const customers = require('./customers')
 const logs = require('./logs')
 const compliance = require('./compliance')
+const notifier = require('./notifier/index')
```
```suggestion
const notifier = require('./notifier')
```

chaotixkilla commented on new-lamassu-admin/src/components/NotificationCenter/NotificationCenter.js:
```diff
@@ -0,0 +1,143 @@
+
+  const handleClearNotification = id => {
+    clearNotification({ variables: { id } })
+  }
+  const buildNotifications = () => {
```
Add newline

```suggestion
  }
  
  const buildNotifications = () => {
```

chaotixkilla commented on new-lamassu-admin/src/components/layout/Header.js:
```diff
@@ -56,8 +70,31 @@ const Header = memo(({ tree }) => {
     history.push('/maintenance/machine-status', { id: machine.deviceId })
   }
 
+  // these inline styles prevent scroll bubbling: when the user reaches the bottom of the notifications list and keeps scrolling,
+  // the body scrolls, stealing the focus from the notification center, preventing the admin from scrolling the notifications back up
+  // on the first scroll, needing to move the mouse to recapture the focus on the notification center
+  // it also disables the scrollbars caused by the notification center's background to the right of the page, but keeps the scrolling on the body enabled
```
The same point as a similar note above

chaotixkilla reviewed (changes requested):
Alright, after a more thorough review this time, a couple issues here and there, should be easy to fix

RafaelTaranto commented on lib/cash-out/cash-out-atomic.js:
```diff
@@ -11,12 +11,12 @@ const cashOutLow = require('./cash-out-low')
 
 const toObj = helper.toObj
 
-module.exports = {atomic}
```
Seems to be only formatting changes, can you undo this?

RafaelTaranto commented on lib/customers.js:
```diff
@@ -115,12 +116,21 @@ async function updateCustomer (id, data, userToken) {
+    return notifierQueries.invalidateNotification(detailB, 'compliance').catch(console.error)
+  }
+  return null
+}
```
if should be separated from the `(`
Also:

```
if (data.authorized_override !== 'verified') return Promise.resolve()

const detailB = notifierUtils.buildDetail({ code: 'BLOCKED', customerId: id })
return notifierQueries.invalidateNotification(detailB, 'compliance')
```

That way the function always returns a promise. Add the catch to the function call (line 119)

RafaelTaranto commented on lib/new-admin/graphql/schema.js:
```diff
@@ -329,6 +344,8 @@ const resolvers = {
     blacklist: () => blacklist.getBlacklist(),
     // userTokens: () => tokenManager.getTokenList()
+    notifications: () => notifierQueries.getNotificationsGql(),
+    hasUnreadNotifications: () => notifierQueries.hasUnreadNotificationsGql()
```
I don't think we need the gql suffix. The functions still work outside the graphql context.

RafaelTaranto commented on lib/new-admin/graphql/schema.js:
```diff
@@ -351,6 +368,8 @@ const resolvers = {
       blacklist.insertIntoBlacklist(cryptoCode, address),
     // revokeToken: (...[, { token }]) => tokenManager.revokeToken(token)
+    clearNotification: (...[, { id }]) => notifierQueries.markAsReadGql(id),
+    clearAllNotifications: () => notifierQueries.markAllAsReadGql()
```
Same as above

RafaelTaranto commented on new-lamassu-admin/src/pages/Notifications/sections/CryptoBalanceOverrides.js:
```diff
@@ -9,8 +9,8 @@ import { transformNumber } from 'src/utils/number'
 
-const HIGH_BALANCE_KEY = 'highBalance'
-const LOW_BALANCE_KEY = 'lowBalance'
+const HIGH_BALANCE_KEY = 'cryptoHighBalance'
+const LOW_BALANCE_KEY = 'cryptoLowBalance'
```
Any reason for this change? It's not really backwards compatible since we already have ops running on the beta version. We would have to run a custom migration just to change the label here.

hellypi replied:
There was a bug of accessing an undefined variable of an object somewhere else in the code. I'll see how I can change the variable names back to the old values

RafaelTaranto commented on new-lamassu-admin/src/pages/Customers/helper.js:
```diff
@@ -11,10 +11,9 @@ const getAuthorizedStatus = it =>
-  return phoneNumber ? phoneNumber.formatInternational() : phone
+  return phone && country
+    ? parsePhoneNumberFromString(phone, country).formatInternational()
+    : ''
```
This will break the function.

RafaelTaranto commented on new-lamassu-admin/src/pages/Customers/components/CustomerDetails.js:
```diff
@@ -9,20 +9,21 @@ import { ReactComponent as LawIconInverse } from 'src/styling/icons/circle butto
 const CustomerDetails = memo(({ customer, locale, setShowCompliance }) => {
   const classes = useStyles()
 
   const elements = [
```
undo all of these changes. This is already fixed on dev.

RafaelTaranto commented on new-lamassu-admin/src/pages/Customers/CustomersList.js:
```diff
@@ -11,7 +11,11 @@ import { ReactComponent as TxOutIcon } from 'src/styling/icons/direction/cash-ou
 import styles from './CustomersList.styles'
-import { getAuthorizedStatus, getFormattedPhone, getName } from './helper'
+import {
+  getAuthorizedStatus,
```
Same thing. This file is not part of the PR. Even in the case of this happening in the future you should keep the changes local.

RafaelTaranto commented on new-lamassu-admin/src/components/typography/styles.js:
```diff
@@ -15,7 +15,7 @@ const base = {
   color: fontColor
 }
 
-export default {
```
Undo, unrelated to PR.

RafaelTaranto commented on new-lamassu-admin/src/components/layout/Header.js:
```diff
@@ -85,14 +122,41 @@ const Header = memo(({ tree }) => {
+                InverseIcon={AddIconReverse}
+                onClick={() => setOpen(true)}>
+                Add machine
+              </ActionButton>
```
Does add machine need to be inside the clickaway listener?

RafaelTaranto commented on new-lamassu-admin/src/components/layout/Header.js:
```diff
@@ -56,8 +70,31 @@ const Header = memo(({ tree }) => {
+  // the body scrolls, stealing the focus from the notification center, preventing the admin from scrolling the notifications back up
+  // on the first scroll, needing to move the mouse to recapture the focus on the notification center
+  // it also disables the scrollbars caused by the notification center's background to the right of the page, but keeps the scrolling on the body enabled
+  const onClickAway = () => {
+    setAnchorEl(null)
+    document.querySelector('body').removeAttribute('style')
+    document.querySelector('#root').removeAttribute('style')
+  }
+
+  const handleClick = event => {
+    setAnchorEl(anchorEl ? null : event.currentTarget)
+    document.querySelector('body').style.overflow = anchorEl ? 'auto' : 'hidden'
+    const root = document.querySelector('#root')
+    root.style.overflowY = 'scroll'
+    root.style.position = 'absolute'
+    root.style.top = '0'
+    root.style.bottom = '0'
+    root.style.left = '0'
+  }
```
Damn, and they're not even handled by react. Here's something you can do:

1- create a class in src/styling/global
2- add and remove the class from the places you need.

RafaelTaranto commented on new-lamassu-admin/src/components/NotificationCenter/NotificationCenter.styles.js:
```diff
@@ -0,0 +1,130 @@
+    overflowX: 'hidden',
+    backgroundColor: white,
+    zIndex: 10,
+    '&::-webkit-scrollbar': {
+      width: 7
+    },
+    '&::-webkit-scrollbar-thumb': {
+      backgroundColor: offColor,
+      borderRadius: 5
+    },
+    '&::-webkit-scrollbar-track-piece:end': {
+      background: 'transparent'
+    }
```
did you test this in other browsers? Scrollbar is usually a PITA to customize.

hellypi replied:
it works on chrome but not on firefox. Should I remove this?

RafaelTaranto commented on migrations/1607009558538-create-notifications-table.js:
```diff
@@ -0,0 +1,40 @@
+var db = require('./db')
+
+function singleQuotify(item) {
+  return "'" + item + "'"
+}
```
```suggestion
const singleQuotify = (item) => `'${item}'`
```

RafaelTaranto commented on migrations/1607009558538-create-notifications-table.js:
```diff
@@ -0,0 +1,40 @@
+  const sql = [
+    `
+    CREATE TYPE notification_type AS ENUM ${'(' + types + ')'};
+    CREATE TABLE IF NOT EXISTS "notifications" (
```
remove the "if not exists", we need to fail hard if the database is in a bad state.

RafaelTaranto commented on lib/plugins.js:
```diff
@@ -1,4 +1,3 @@
-const uuid = require('uuid')
```
this filie is a complete mess because of the automatic formatter. undo all changes unrelated to the functionality

RafaelTaranto commented on lib/notifier/test/utils.test.js:
```diff
@@ -0,0 +1,100 @@
+  ).toBe('string')
+})
+
+test('Send no alerts returns empty object with sms and email disabled', () => {
+  expect(utils.sendNoAlerts(plugins, false, false)).toEqual({})
+})
+
+test('Send no alerts returns object with sms prop with sms only enabled', () => {
+  expect(utils.sendNoAlerts(plugins, true, false)).toEqual({
+    sms: {
+      body: '[Lamassu] All clear'
+    }
+  })
+})
+
+test('Send no alerts returns object with sms and email prop with both enabled', () => {
+  expect(utils.sendNoAlerts(plugins, true, true)).toEqual({
+    email: {
+      body: 'No errors are reported for your machines.',
+      subject: '[Lamassu] All clear'
+    },
+    sms: {
+      body: '[Lamassu] All clear'
+    }
+  })
+})
+
+test('Send no alerts returns object with email prop if only email is enabled', () => {
+  expect(utils.sendNoAlerts(plugins, false, true)).toEqual({
+    email: {
+      body: 'No errors are reported for your machines.',
+      subject: '[Lamassu] All clear'
+    }
+  })
+})
```
This is a group of tests, here's what you can do:

```
describe('sendNoAlerts', () => {
  test('Should return only email', () => {})
  test('Should return only sms', () => {})
})
```

This can be applied to a bunch of tests on this PR.

RafaelTaranto reviewed (changes requested):
Still need to review the main files.

RafaelTaranto commented on lib/config-migration.js:
```diff
@@ -187,7 +187,7 @@ function migrateNotifications (config) {
     cashOutCassette1AlertThreshold: 'fiatBalanceCassette1',
     cashOutCassette2AlertThreshold: 'fiatBalanceCassette2',
-    cryptoAlertThreshold: 'cryptoLowBalance'
+    cryptoAlertThreshold: 'lowBalance'
```
This won't really work because we already have ops running on the beta.

RafaelTaranto commented on new-lamassu-admin/src/pages/Notifications/sections/CryptoBalanceAlerts.js:
```diff
@@ -6,8 +6,8 @@ import SingleFieldEditableNumber from '../components/SingleFieldEditableNumber'
-const HIGH_BALANCE_KEY = 'cryptoHighBalance'
+const LOW_BALANCE_KEY = 'lowBalance'
+const HIGH_BALANCE_KEY = 'highBalance'
 
```
Same as above, ops running the beta will "lose" this value in their config files.

RafaelTaranto commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,384 @@
+    if (res.length > 0) return Promise.resolve()
+    return queries.addNotification(COMPLIANCE, message, detailB)
+  }).catch(console.error)
+}
```
formatting 
```suggestion
  return clearOldCustomerSuspendedNotifications(customer.id, deviceId)
    .then(() => queries.getValidNotifications(COMPLIANCE, detailB))
    .then(res => {
      if (res.length > 0) return Promise.resolve()
      return queries.addNotification(COMPLIANCE, message, detailB)
    })
    .catch(console.error)
```

RafaelTaranto commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,384 @@
+  } else {
+    detailB = utils.buildDetail({ cryptoCode: tx.cryptoCode, code: 'BLOCKED', cryptoAddress: tx.toAddress })
+    message = `Blocked blacklisted address: ${tx.cryptoCode} ${tx.toAddress.substr(0, 10)}...`
+  }
```
```suggestion
  const code = isAddressReuse ? 'REUSED' : 'BLOCKED'
  const name = isAddressReuse ? 'reused' : 'blacklisted'
  
  const detailB = utils.buildDetail({ cryptoCode: tx.cryptoCode, code, cryptoAddress: tx.toAddress })
  const message = `Blocked ${name} address: ${tx.cryptoCode} ${tx.toAddress.substr(0, 10)}...`
```

RafaelTaranto commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,384 @@
+          const detailB = utils.buildDetail({ code: PING, age: alert.age ? alert.age : -1, deviceId: alert.deviceId })
+          return queries.getValidNotifications(ERROR, _.omit(['age'], detailB)).then(res => {
+            if (res.length > 0) return Promise.resolve()
+            console.log('Adding PING alert on database for ' + alert.machineName)
```
Remove this console.log

RafaelTaranto commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,384 @@
+          const detailB = utils.buildDetail({ code: STALE, deviceId: alert.deviceId })
+          return queries.getValidNotifications(ERROR, detailB).then(res => {
+            if (res.length > 0) return Promise.resolve()
+            console.log('Adding STALE alert on database for ' + alert.machineName)
```
same as above

RafaelTaranto commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,384 @@
+    }, alertRec.devices[device].deviceAlerts)
+    // concat every array into one
+    alerts = _.concat(alerts, alertRec.devices[device].deviceAlerts)
+  })
```
```suggestion
const embedDeviceId = deviceId => _.assign({ deviceId })

const mapToAlerts = _.map(it => _.map(embedDeviceId(it), alertRec.devices[it].deviceAlerts))
const alerts = _.compose(_.flatten, mapToAlerts, _.keys)(alertRec.devices)
```

RafaelTaranto commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,384 @@
+    }, res)
+    return indexesToInvalidate.length > 0 ? queries.batchInvalidate(indexesToInvalidate) : null
+  }).catch(console.log)
+}
```
```suggestion
const clearOldErrorNotifications = alerts => {
  return queries.getAllValidNotifications(ERROR)
    .then(res => {
      // for each valid notification in DB see if it exists in alerts
      // if the notification doesn't exist in alerts, it is not valid anymore
      const filterByAlert = _.filter(notification => {
        const { code, deviceId } = notification.detail
        return !_.find(alert => alert.code === code && alert.deviceId === deviceId)(alerts)
      })

      const indexesToInvalidate = _.compose(_.map('id'), filterByAlert)(res)

      if (!indexesToInvalidate.length) return Promise.resolve()
      
      return queries.batchInvalidate(indexesToInvalidate)
    })
    .catch(console.log)
}
```

RafaelTaranto commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,384 @@
+        // if it exists then it is a duplicate, add it to array
+        duplicateIndexes.push(idx)
+      }
+    }, notifications)
```
This can be refactored to work similarly as the other one

RafaelTaranto commented on lib/notifier/index.js:
```diff
@@ -0,0 +1,384 @@
+        duplicateIndexes.push(idx)
+      }
+    }, notifications)
+    return (idsToInvalidate.length > 0 ? queries.batchInvalidate(idsToInvalidate) : Promise.resolve()).then(() => duplicateIndexes)
```
This is kinda of a mess. If there's IDs to invalidate we don't return duplicateIndexes, is there a reason for that?

hellypi replied:
it's to prevent readding an existing notification, that wasn't invalidated. It could use a better name and implementation anyway, so I fixed it

RafaelTaranto commented on lib/routes.js:
```diff
@@ -358,7 +361,10 @@ function triggerSuspend (req, res, next) {
   const date = new Date()
   date.setDate(date.getDate() + days);
   customers.update(id, { suspendedUntil: date })
-    .then(customer => respond(req, res, { customer }))
+    .then(customer => {
+      notifier.customerComplianceNotify(customer, req.deviceId, 'SUSPENDED', days)
+      respond(req, res, { customer })
+    })
```
this changes behaviour, you need to return the `respond(` call

RafaelTaranto commented on lib/plugins.js:
```diff
@@ -686,24 +603,26 @@ function plugins (settings, deviceId) {
-    const highAlertThreshold = notifications.cryptoHighBalance
+    const override = _.find(override => override.cryptoCurrency === cryptoCode, settings.config.notifications_cryptoBalanceOverrides)
+    const lowAlertThreshold = override ? override.lowBalance : notifications.cryptoLowBalance
+    const highAlertThreshold = override ? override.highBalance : notifications.cryptoHighBalance
```
I'm pretty sure that this code is redundant, getNotifications passing down cryptoCode should already be resolving the overrides for you.

hellypi committed:
Fix: fix crypto overrides not being respected

RafaelTaranto commented on lib/new-config-manager.js:
```diff
@@ -78,17 +70,23 @@ const getAllCryptoCurrencies = (config) => {
+
+  const cryptoOverride = _.filter(o => o.cryptoCurrency === cryptoCurrency)(notifications.cryptoBalanceOverrides)[0]
+  const cryptoHighBalance = _.getOr(notifications.cryptoHighBalance, ['highBalance'], cryptoOverride) // cryptoOverride ? cryptoOverride.highBalance : notifications.cryptoHighBalance
+  const cryptoLowBalance = _.getOr(notifications.cryptoLowBalance, ['lowBalance'], cryptoOverride) // cryptoOverride ? cryptoOverride.lowBalance : notifications.cryptoLowBalance
+  const cryptoSettings = { cryptoHighBalance, cryptoLowBalance }
 
-  const fiatFilter = _.matches({ machine })
-  const withFiatBalance = resolveOverrides(withCryptoBalance, fiatFilter, withCryptoBalance.fiatBalanceOverrides, 'fiatBalanceOverrides')
+  const fiatOverride = _.filter(o => o.machine === machine)(notifications.fiatBalanceOverrides)[0]
+  const fiatBalanceCassette1 = _.getOr(notifications.fiatBalanceCassette1, ['fiatBalanceCassette1'], fiatOverride)
+  const fiatBalanceCassette2 = _.getOr(notifications.fiatBalanceCassette2, ['fiatBalanceCassette2'], fiatOverride)
+  const fiatSettings = { fiatBalanceCassette1, fiatBalanceCassette2 }
```
I think you can simplify this with:

```js
const findByCryptoCurrency = _.find(_.matches({ cryptoCurrency }))
const findByMachine = _.find(_.matches({ machine }))

const cryptoFields = ['cryptoHighBalance', 'cryptoLowBalance']
const fiatFields= ['fiatBalanceCassette1', 'fiatBalanceCassette2']

const getCryptoSettings = _.compose(_.pick(cryptoFields), _.defaultTo(notifications), findByCryptoCurrency)
const cryptoSettings = getCryptoSettings(notifications.cryptoBalanceOverrides)

const getFiatSettings = _.compose(_.pick(fiatFields), _.defaultTo(notifications), findByMachine)
const fiatSettings = getFiatSettings(notifications.fiatBalanceOverrides)
```

RafaelTaranto commented on lib/new-config-manager.js:
```diff
@@ -78,17 +70,23 @@ const getAllCryptoCurrencies = (config) => {
 
-  const withSms = fromNamespace('sms', withFiatBalance)
-  const withEmail = fromNamespace('email', withFiatBalance)
+  const id = cryptoOverride ? cryptoOverride.id : fiatOverride ? fiatOverride.id : null
```
Hmm, I know I told this is important to preserve, but this is only preserving for cryptoOverrides, not fiatOverrides. You can take this out, we can export when needed.

RafaelTaranto commented on lib/new-config-manager.js:
```diff
@@ -78,17 +70,23 @@ const getAllCryptoCurrencies = (config) => {
 
-  const final = { ...withFiatBalance, sms: withSms, email: withEmail }
-  return final
+  return { ...notifNoOverrides, sms: smsSettings, email: emailSettings, ...cryptoSettings, ...fiatSettings, machine, cryptoCurrency, id }
```
I don't think we need to re-export machine or crypto, code that calls this function already has that info.
